Userland/Libraries/LibWeb/Painting/RecordingPainter.cpp

```diff
@@ -0,0 +1,829 @@
+    painter.translate(translation_delta);
+}
+
+void SaveState::execute(CommandExecutionState& state) const
+{
+    auto& painter = state.painter();
+    painter.save();
+}
+
+void RestoreState::execute(CommandExecutionState& state) const
+{
+    auto& painter = state.painter();
+    painter.restore();
+}
+
+void AddClipRect::execute(CommandExecutionState& state) const
+{
+    auto& painter = state.painter();
+    painter.add_clip_rect(rect);
+}
+
+void ClearClipRect::execute(CommandExecutionState& state) const
+{
+    auto& painter = state.painter();
+    painter.clear_clip_rect();
+}
+
+void SetFont::execute(CommandExecutionState& state) const
+{
+    auto& painter = state.painter();
+    painter.set_font(font);
+}
+
+void PushStackingContext::execute(CommandExecutionState& state) const
+{
+    auto& painter = state.painter();
+    if (has_fixed_position) {
+        painter.translate(-painter.translation());
+    }
+    if (semitransparent_or_has_non_identity_transform) {
+        auto destination_rect = transformed_destination_rect.to_rounded<int>();
+
+        // FIXME: We should find a way to scale the paintable, rather than paint into a separate bitmap,
+        // then scale it. This snippet now copies the background at the destination, then scales it down/up
+        // to the size of the source (which could add some artefacts, though just scaling the bitmap already does that).
+        // We need to copy the background at the destination because a bunch of our rendering effects now rely on
+        // being able to sample the painter (see border radii, shadows, filters, etc).
+        Gfx::FloatPoint destination_clipped_fixup {};
+        auto try_get_scaled_destination_bitmap = [&]() -> ErrorOr<NonnullRefPtr<Gfx::Bitmap>> {
+            Gfx::IntRect actual_destination_rect;
+            auto bitmap = TRY(painter.get_region_bitmap(destination_rect, Gfx::BitmapFormat::BGRA8888, actual_destination_rect));
+            // get_region_bitmap() may clip to a smaller region if the requested rect goes outside the painter, so we need to account for that.
+            destination_clipped_fixup = Gfx::FloatPoint { destination_rect.location() - actual_destination_rect.location() };
+            destination_rect = actual_destination_rect;
+            if (source_rect.size() != transformed_destination_rect.size()) {
+                auto sx = static_cast<float>(source_rect.width()) / transformed_destination_rect.width();
+                auto sy = static_cast<float>(source_rect.height()) / transformed_destination_rect.height();
+                bitmap = TRY(bitmap->scaled(sx, sy));
+                destination_clipped_fixup.scale_by(sx, sy);
+            }
+            return bitmap;
+        };
+
+        auto bitmap_or_error = try_get_scaled_destination_bitmap();
+        if (bitmap_or_error.is_error()) {
+            // NOTE: If creation of bitmap failed we need to skip all painting commands that belong to this stacking context.
```

Comment:
I think it would be okay if we require that allocating the bitmap succeeds here.
If it fails, we will not render the page correctly, and there is no in-spec mechanism for recovering and/or reporting this kind of problem.

If we make this temporary bitmap allocation infallible, it simplifies the display list iteration and makes it less confusing how traversal works :)

Reply:
Actually `get_region_bitmap()` has a bit confusing api and returns an error if requested bitmap is located outside of viewport (has 0x0 size) so we should proceed executing painting commands in that case. I added a FIXME that explains that.

For now I changed all commands to return `Continue` or `SkipStackingContext` and let main commands execution loop handle skipping. I hope it makes `PushStackingContext` less confusing :)